onnx/inliner/inliner.cc

```diff
@@ -262,11 +310,120 @@ class Specializer {
   }
 };
 
-void InlineFunctions(ModelProto& model, FunctionResolver resolver) {
+// Identify the set of all "input" variables used by a given node.
+// This includes the variables listed as node.input, as well as
+// implicit inputs referred to in any graph-valued-attribute of the node.
+
+struct ComputeUsedVars : public NodeVisitorBase {
+  std::vector<std::string> result;
+
+  bool ProcessNode(const NodeProto& node) override {
+    for (auto& var : node.input()) {
+      if (!var.empty()) {
+        result.push_back(var);
+      }
+    }
+    return true; // process sub-graphs
+  }
+
+  ComputeUsedVars(const NodeProto& node) {
+    result.reserve(node.input_size());
+    VisitNode(node);
+  }
+};
+
+std::vector<std::string> GetUsedVars(const NodeProto& node) {
+  return ComputeUsedVars(node).result;
+}
+
+using ConstNodeMap = std::unordered_map<std::string, const NodeProto*>;
+
+ConstNodeMap FindConstantNodes(const GraphProto& graph) {
+  ConstNodeMap result;
+  for (const NodeProto& node : graph.node()) {
+    if (IsOnnxDomain(node.domain()) && (node.op_type() == "Constant")) {
+      result[node.output(0)] = &node;
+    }
+  }
+  return result;
+}
+
+const TypeProto& GetType(const ModelProto& model, std::string var) {
+  for (auto& vi : model.graph().value_info()) {
+    if (vi.name() == var)
+      return vi.type();
+  }
+  for (auto& vi : model.graph().input()) {
+    if (vi.name() == var)
+      return vi.type();
+  }
+  for (auto& vi : model.graph().output()) {
+    if (vi.name() == var)
+      return vi.type();
+  }
+  ONNX_ASSERTM(false, "Type unknown for ", var);
+}
+
+void ConvertVersion(ModelProto& model, const NodeProto& call_node, FunctionProto& function, int target_version) {
```

Comment:
What if there are multiple opset involved?

Reply:
Currently, the version-converter supports only version-conversion for a single domain. If and when the version-converter is extended to support other domains, we can extend this too.

Reply:
another PR i guess